### ISSUE TITLE
warning: download buffer is full; consider increasing the 'download-buffer-size' setting (won't work)

### ISSUE THREAD

**Issue by roberth** [NixOS/nix#11728]
**Describe the bug**

I'm getting this warning during substitution and/or remote builds.
These machines have gigabit connections.

I'd need buffers of hundreds of megabytes to avoid this warning when the store paths are big.
(Now imagine a well connected small machine without swap)

**Steps To Reproduce**

`nix-build -A nixosTests.nixos-rebuild-target-host`

after a GC or on a fresh "generation" of Nixpkgs

**Expected behavior**

When the download buffer is full, continue downloading at the rate of the consumer, and don't complain about it.

_Limit_ the buffer size to avoid jittery consumption, which is known to cause connections to drop when the pauses are too long.
I've already outlined this as a better solution in the PR that introduced this problem,
- https://github.com/NixOS/nix/pull/11171

**`nix-env --version` output**

**Additional context**

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by roberth** [comment]
Also occurs when copying between local stores, with `nix store repair --extra-substituters /some/path?trusted=true` for example; probably also `nix copy --to /some/path` etc.

**Comment by malikwirin** [comment]
So what is the solution to no longer get this warning?

**Comment by boozedog** [comment]
i'm getting this too

**Comment by l0b0** [comment]
> So what is the solution to no longer get this warning?

Looks like setting [`download-buffer-size`](https://nix.dev/manual/nix/2.25/command-ref/conf-file.html#conf-download-buffer-size) bigger than the current default of `67108864` (65536 kibibytes??)

**Comment by imincik** [comment]
Just hit this issue as well on my machine which is using the remote builder

```
buffer is full; consider increasing the 'download-buffer-size' setting
warning: download buffer is full; consider increasing the 'download-buffer-size' setting
```

**Comment by apholden** [comment]
> This issue has been mentioned on **NixOS Discourse**. There might be relevant details there:
> 
> https://discourse.nixos.org/t/2024-12-13-nix-team-meeting-minutes-202/57281/1

I just got this for the first time while running nix-shell.  I followed the link to NixOS Discourse, and all I found was a link back to this issue.  Could someone explain what I did wrong?

**Comment by davychhouk** [comment]
> > So what is the solution to no longer get this warning?
> 
> Looks like [the default used to be 1 mebibyte until Nix 2.24](https://github.com/NixOS/nix/pull/11171/files#diff-ac4189bb1b78aa9d7e8f334ca622f6dc25da6a3f128c1c9475aad6bf220a000dL854). Since then you can also set [`download-buffer-size`](https://nix.dev/manual/nix/2.25/command-ref/conf-file.html#conf-download-buffer-size). That said, I still see this warning in CI with the current default of `67108864` (64 mebibytes).
> 
> [Bumping to 500 MB](https://gitlab.com/engmark/root/-/merge_requests/785/diffs) seems to work for my builds.

Yep, it seems like it's going away after bumping it to `500MB`.
You could do this by adding nix config in your `configuration.nix` file:

`nix.settings.download-buffer-size = 524288000;`

Then if you apply, you should be able to see the `nix.conf` (It should be at `/etc/nix/nix.conf`) updated with the `download-buffer-size` set to `524288000`.

Then there should be no warning anymore.

**Comment by ArkieSoft** [comment]
> > > So what is the solution to no longer get this warning?
> > 
> > 
> > Looks like [the default used to be 1 mebibyte until Nix 2.24](https://github.com/NixOS/nix/pull/11171/files#diff-ac4189bb1b78aa9d7e8f334ca622f6dc25da6a3f128c1c9475aad6bf220a000dL854). Since then you can also set [`download-buffer-size`](https://nix.dev/manual/nix/2.25/command-ref/conf-file.html#conf-download-buffer-size). That said, I still see this warning in CI with the current default of `67108864` (64 mebibytes).
> > [Bumping to 500 MB](https://gitlab.com/engmark/root/-/merge_requests/785/diffs) seems to work for my builds.
> 
> Yep, it seems like it's going away after bumping it to `500MB`. You could do this by adding nix config in your `configuration.nix` file:
> 
> `nix.settings.download-buffer-size = 524288000;`
> 
> Then if you apply, you should be able to see the `nix.conf` (It should be at `/etc/nix/nix.conf`) updated with the `download-buffer-size` set to `524288000`.
> 
> Then there should be no warning anymore.

I am now getting this issue, but there is no nix option [nix.settings.download-buffer-size](https://search.nixos.org/options?channel=25.05&from=0&size=50&sort=relevance&type=packages&query=nix.settings.download-buffer-size) in 25.05, and oddly not even in 24.11. It's still in the documentation to exist for the [documentation](https://nix.dev/manual/nix/2.28/command-ref/conf-file.html?highlight=download-buffer-size#conf-download-buffer-size) for 2.28 (standard in NixOS 25.05). Any clues?

**Comment by roberth** [comment]
Not all settings are explicitly listed in the NixOS options, but all of them are available nonetheless when you have them in your config.

This is a UX issue with the options search:
- https://github.com/NixOS/nixos-search/issues/959

**Comment by davychhouk** [comment]
> > > > So what is the solution to no longer get this warning?
> > > 
> > > 
> > > Looks like [the default used to be 1 mebibyte until Nix 2.24](https://github.com/NixOS/nix/pull/11171/files#diff-ac4189bb1b78aa9d7e8f334ca622f6dc25da6a3f128c1c9475aad6bf220a000dL854). Since then you can also set [`download-buffer-size`](https://nix.dev/manual/nix/2.25/command-ref/conf-file.html#conf-download-buffer-size). That said, I still see this warning in CI with the current default of `67108864` (64 mebibytes).
> > > [Bumping to 500 MB](https://gitlab.com/engmark/root/-/merge_requests/785/diffs) seems to work for my builds.
> > 
> > 
> > Yep, it seems like it's going away after bumping it to `500MB`. You could do this by adding nix config in your `configuration.nix` file:
> > `nix.settings.download-buffer-size = 524288000;`
> > Then if you apply, you should be able to see the `nix.conf` (It should be at `/etc/nix/nix.conf`) updated with the `download-buffer-size` set to `524288000`.
> > Then there should be no warning anymore.
> 
> I am now getting this issue, but there is no nix option [nix.settings.download-buffer-size](https://search.nixos.org/options?channel=25.05&from=0&size=50&sort=relevance&type=packages&query=nix.settings.download-buffer-size) in 25.05, and oddly not even in 24.11. It's still in the documentation to exist for the [documentation](https://nix.dev/manual/nix/2.28/command-ref/conf-file.html?highlight=download-buffer-size#conf-download-buffer-size) for 2.28 (standard in NixOS 25.05). Any clues?

This is not a NixOS option specifically, but it's a nix config option.
You can find nix's doc here: https://nix.dev/manual/nix/2.29/command-ref/conf-file.html#conf-download-buffer-size

**Comment by vorburger** [comment]
Just FYI, I'm still seeing this, on non-NixOS (on Fedora), when I `nix build .#livecd` [here](https://github.com/vorburger/vorburger-dotfiles-bin-etc/tree/main/NixOS), AFTER setting [`download-buffer-size = 524288000`](https://nix.dev/manual/nix/2.28/command-ref/conf-file#conf-download-buffer-size) (500 MB) in my [` ~/.config/nix/nix.conf`](https://github.com/vorburger/vorburger-dotfiles-bin-etc/blob/main/dotfiles/nix/nix.conf) AND `trusted-users` in [`/etc/nix/nix.custom.conf`](https://github.com/vorburger/vorburger-dotfiles-bin-etc/blob/main/etc/nix/nix.custom.conf). 

I'm on a 940 Mbps fiber, and it sounds like that could be related? I would be cool to have a "formula" for this.

For now, I'll double it and use `download-buffer-size = 1048576000`, and see if I'll still see it in the future.

**Comment by vorburger** [comment]
> For now, I'll double it and use download-buffer-size = 1048576000, and see if I'll still see it in the future.

Still seeing it; but just realized that I forgot to `systemctl restart nix-daemon.service`, did that now; let's see. -- If seeing it again, maybe I'll quadruple it (so x8 from original suggestion here) to `4194304000` (4 GB), and report back if still seeing it. But if so, I suspect that it's more likely that change / setting is somehow just not being used, for some reason.

**Comment by xokdvium** [comment]
So the underlying issue here is that nix is very bad at controlling backpressure from libcurl and doesn't pause transfers that are in-flight if it can't consume the data fast enough like in the case of decompressing lzma, which is notoriously slow. There are some pretty trivial things we could do to further optimize this.

**Comment by alex-robbins** [comment]
It sounds like increasing the download buffer size is not actually a good way to handle this situation, even if it does make the warning go away in some cases. If that's true, then I think the warning should be removed now, even while the underlying issue (about controlling backpressure) is still being worked on/tracked here. (That way people wouldn't be blindly increasing the download-buffer-size setting. It sounds like there's a lot of confusion/distraction in this thread because of that.)

**Comment by xokdvium** [comment]
> download buffer size is not actually a good way to handle this situation

Yes, that only cures the symptom of the issue that can only be fixed in nix. But also hitting the download buffer size limit is very bad for download speed since it stalls the whole download thread and prevents multiplexing. The correct solution is to implement pausing in nix itself. I hope to get started on improving the situation here soon-ish.

**Comment by xokdvium** [comment]
For anybody watching this issue, I've put up https://github.com/NixOS/nix/pull/14614. Testing and benchmarking that patch out would be greatly appreciated :)

**Comment by Diti** [comment]
I’m getting that error too. Currently on 8 Gbps fiber (I have SFP+ NICs on my computers, if that’s relevant info). What size should the buffer be?

**Comment by xokdvium** [comment]
This has been fixed in >= 2.33. Please see if newer version improves the situation for you.

**Comment by BerriJ** [comment]
Do you know when this version will hit the nixos-unstable channel? It's still on 2.31.3 and I can't find a PR that will update it.
I also find the derivation confusing so I don't really know how to open up that PR myself.

**Comment by ilkecan** [comment]
@BerriJ It seems 2.33 is added to `nixos-unstable` 40 days ago (https://github.com/NixOS/nixpkgs/commit/959699cac37d690fd6cb5970d7c508b618bb835a).

You can access different versions of Nix through (`pkgs.nixVersions`):
```nix
nix-repl> pkgs.unstable.nixVersions.nix_2_33
«derivation /nix/store/prf1ckynra9n68qb251slpn8x3cfkcc8-nix-2.33.3.drv»

nix-repl> pkgs.unstable.nixVersions.latest
«derivation /nix/store/prf1ckynra9n68qb251slpn8x3cfkcc8-nix-2.33.3.drv»
```

**Comment by BerriJ** [comment]
> [@BerriJ](https://github.com/BerriJ) It seems 2.33 is added to `nixos-unstable` 40 days ago ([NixOS/nixpkgs@959699c](https://github.com/NixOS/nixpkgs/commit/959699cac37d690fd6cb5970d7c508b618bb835a)).
> 
> You can access different versions of Nix through (`pkgs.nixVersions`):
> 
> nix-repl> pkgs.unstable.nixVersions.nix_2_33
> «derivation /nix/store/prf1ckynra9n68qb251slpn8x3cfkcc8-nix-2.33.3.drv»
> 
> nix-repl> pkgs.unstable.nixVersions.latest
> «derivation /nix/store/prf1ckynra9n68qb251slpn8x3cfkcc8-nix-2.33.3.drv»

I see that PR and I agree its part of the master and nixos-unstable branches now, but on my up-to-date flake based system (following nixos-unstable) I get:

```sh
❯ nix --version
nix (Nix) 2.31.3
```
`nix run` also shows 2.31.3:

```sh
nix run nixpkgs/nixos-unstable#nix -- --version
nix (Nix) 2.31.3
```

and my /nix/store also does not contain 3.33

```sh
❯ ls /nix/store | grep nix-2.3
1m15g16avpxksw4g0vl3ihmn7wd3xpii-nix-2.31.2+1.drv
26ry38cqp1mgas5bnn4204i2nfxhrsw1-nix-2.31.2+1
2imns8c6aiyxrnhwani3srv9ygk0d65f-nix-2.31.3
2wgak0m2iaga21fyvgclxs9ciyy378pb-nix-2.31.2+1.drv
2yv30mfayi5z9dfd5mm6z5hfqij6q8kb-nix-2.31.3.drv
3prabirag8wy703f0hhhrjbj3x3kdjx6-nix-2.31-14240-sri-error-message.patch.drv
5vsq7w2pn1pd5z3f4917akykycn04wnw-nix-2.30-14695-mdbook-0.5-support.patch.drv
64dgnadlpiw1dbqigpnv1icp7shb8c18-nix-2.31.3-doc
6gzykc5bg094n1jjfxdgwmh5wz0zbh38-nix-2.31.3
6hxjfn55qd0yj85wh82rb0xb35jk8sjn-nix-2.31.2+1.drv
7bnvywyq8bymzcnlziqx2k5hw4969ilb-nix-2.31.2+1
96lwmm2bsyasw9mk7rhzk1avmc56wdzl-nix-2.31.2+1-man
9c9akczfyrsynz5gc0f2kwz2wqs9mczx-nix-2.31.2+1-doc
9mpgjc4nnyvg78vck6wm3r7a3q8rq44j-nix-2.30-14695-mdbook-0.5-support.patch.drv
bad4a07031nnsv6qschda5pwas43006c-nix-2.31.3.drv
ci3b1vhkbdcpqxjw7npxfgldwap7ij04-nix-2.31.3.drv
dv7b4i51k7m66h5k6y1qk7gipg33dbf8-nix-2.31.3.drv
fxb5z2rcxmxwh81am5xs32569vmcx0cz-nix-2.31.2+1.drv
h5vc496ffmx75nmvpinar8bxd2ryf7pm-nix-2.31.3.drv
hvxz2mxzzr2aqqvdwqcdlrfkg2mwnsgl-nix-2.31.3-doc
i5k130pbirxlpva6mqz06v4nh7lb452y-nix-2.31.2+1
ijz4jl7wqb6s3xi2vlvybv1kk592ci6j-nix-2.31-14240-sri-error-message.patch.drv
j2nm6lmfzrxzyhira6i5kmrp4j3zsn5g-nix-2.31.3-man
jbbrd9aimadwv8cladqikihf9lb9dl2y-nix-2.31.3-man
jifpj10jncryriff4nn79xvzihcvqvw0-nix-2.31.3.drv
jllly8vb30kw8fy2y18xydnx6nh51407-nix-2.31-14240-sri-error-message.patch.drv
jsrv1x7p0d64lnq4mhszmxzndk9xpn19-nix-2.31.3.drv
k34y7hamyw7xqj22w1a7h1hfi2xdzvk1-nix-2.31.2+1.drv
kkn8nlj8ps9x1qfms4kns9gwk5d93cmb-nix-2.31.3-doc
lwrzv2k3l7wq4q3mzcvrx1qp9jn0r38l-nix-2.31.3.drv
ly5i2vg0p0w8pj5zsqrzqzq3y8jg3kaw-nix-2.31.2+1
qrvg42qrrsk1ism0h6pnma1bmvdqzdl8-nix-2.31.3
r939fjxxzjx82msqnjr1qq13wdb290dg-nix-2.31.2+1-doc
rc3bc0q4ibhqv2rkwi2yp0b8msp5kpk6-nix-2.31.2+1.drv
rlsp5454al1rjkfnz55n04b6cisk4byi-nix-2.31.3-man
skhkr19yq4bxblmpahhixxp0xwv30656-nix-2.31.3.drv
v4h52l5mkba144asciyz05358l3m1h3p-nix-2.31.2+1.drv
vb54x36p7r2lfn82ah45yddvhhl6b94l-nix-2.31.2+1.drv
w5g7r2fw921pl7k2iw4h8mdpi2lsmcr1-nix-2.31-14240-sri-error-message.patch.drv
ws4cmrxc1r0h8nnyb8nj44ix4bz1850m-nix-2.31-14240-sri-error-message.patch.drv
wxm9qds5wayqnfryk3mgqikg0v4gm98l-nix-2.31.3
x6aymss2n7j5xqplg7ahlspb6ax9n4p9-nix-2.31.3.drv
y7m2phpvda41km4m4hqg1p08acvqi44r-nix-2.31.3
yamhjib1kj7a9vs1fna5g6pz0fi2y1l3-nix-2.31.3
ybvz3b2b0ypp4nydi0h8gd0nrxis5wjr-nix-2.31.2+1.drv
zjw7pb2phf0cmqlgh9nxyyqv1dk4jafh-nix-2.31.2+1-man
zmd3mcyrl8wycppjgya34bx51x424h4y-nix-2.31.2+1.drv
```

What am I missing here?

**Comment by xokdvium** [comment]
Default and currently stable nix version is 2.31 - 2.33 addition to 25.11 is awaiting review/merge in https://github.com/NixOS/nixpkgs/pull/491376.

**Comment by BerriJ** [comment]
> Default and currently stable nix version is 2.31 - 2.33 addition to 25.11 is awaiting review/merge in [NixOS/nixpkgs#491376](https://github.com/NixOS/nixpkgs/pull/491376).

I agree on that. But my `nix run` command should use `nixos-unstable` like I specified, right?

`nix run nixpkgs/nixos-unstable#nix -- --version`

**Comment by xokdvium** [comment]
Default on unstable is also 2.31. To use 2.33:

```
nix run nixpkgs/nixos-unstable#nixVersions.nix_2_33 -- --version
```
